Voyageai integration

fzowl opened the pull request:
<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against main? Unless there is a good reason otherwise, we prefer pull requests against main and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

fzowl committed:
VoyageAI embeddings and rerank:
 - embeddings works, tested
 - initial rerank code

What's missing:
 - unit and integration tests
 - rerank request/response mapping and verification

prdoyle committed:
Remove entitlement reflection ITs. (#121550)

This is best tested in unit tests, not REST actions.

smalyshev committed:
Refactor CCS tests (#121547)

* Refactor CCS tests

Clean up duplication, move stop tests to separate test

carsonip committed:
Add missing apm-server tail sampling monitoring metrics to stack monitoring mapping (#121543)

Add missing apm-server tail sampling monitoring metrics to stack monitoring mapping. They were missed in #110568.

Mikep86 committed:
Enable New Semantic Text Format Only On Newly Created Indices (#121556)

nik9000 committed:
ESQL: Reenable heap attacks (#121565)

Reenables some heap attack tests, bumping their memory requirements to
try and force a failure on all CI machines. Previously some CI machines
weren't failing, invalidating the test on those machines.

Close #121481 Close #121465

fzowl committed:
VoyageAI embeddings and rerank:
 - embeddings works, tested
 - rerank works, tested (https://www.elastic.co/search-labs/blog/elasticsearch-cohere-rerank)

What's missing:
 - unit and integration tests

fzowl committed:
VoyageAI embeddings and rerank:
 - embeddings works, tested
 - rerank works, tested (https://www.elastic.co/search-labs/blog/elasticsearch-cohere-rerank)

What's missing:
 - unit and integration tests

fzowl committed:
VoyageAI embeddings and rerank:
 - embeddings works, tested
 - rerank works, tested (https://www.elastic.co/search-labs/blog/elasticsearch-cohere-rerank)

What's missing:
 - unit and integration tests

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.action.CrossClusterQueryUnavailableRemotesIT testRemoteOnlyCCSAgainstDisconnectedRemoteWithSkipUnavailableTrue #121578

ymao1 committed:
Adding patch transport version (#121560)

nik9000 committed:
ESQL: Reenable another heap attack (#121572)

This one was disabled by a long-closed issue.

masseyke committed:
Passing parent task id to the refresh request within reindex data streams (#121533)

lukewhiting committed:
Various Datastream Reindex Fixes (#121376)

* Fix incorrect value of reindex_required flag on ignored index warning
* Datastream reindex now uses unverified write block to allow retrying failed reindex

smalyshev committed:
Implement runtime skip_unavailable=true (#121240)

* Implement runtime skip_unavailable=true

dnhatn committed:
Fix ClusterRequestTests (#121570)

The upper bound of randomVersionBetween is inclusive; therefore, for 
testing the fallback version of the request, we need to use the version
preceding 8.16.0 rather than 8.16.0 itself.

Closes #117937

nielsbauman committed:
Run `TransportEnrichStatsAction` on local node (#121256)

This action solely needs the cluster state, it can run on any node.
Additionally, it needs to be cancellable to avoid doing unnecessary work
after a client failure or timeout.

dnhatn committed:
Wait for exchange source to complete before verifying results (#121603)

Closes #118732

dnhatn committed:
Unwrap failure in testCancelRequestWhenFailingFetchingPages (#121597)

The thrown exception may be wrapped in an ExecutionException; therefore, 
we need to unwrap it before verification.

Closes #121596
Closes #118193

dnhatn committed:
Fix relocation targets in FieldCapabilitiesIT (#121606)

The test failed because we tried to move a shard to a node that already 
has a copy. This change prevents that from happening.

Closes #119280
Closes #120772

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.action.CrossClustersCancellationIT testTasks #121626

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.action.CrossClustersCancellationIT testCloseSkipUnavailable #121627

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.action.CrossClustersCancellationIT testCancel #121632

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.action.CrossClustersCancellationIT testCancelSkipUnavailable #121631

fcofdez committed:
Revert "Remove INDEX_REFRESH_BLOCK after index becomes searchable (#120807)" (#121427)

This reverts commit ae0f1a64b571c319e33a24bc8a05a1fa1d1668b9.

The refresh block would be removed in a subsequent cluster state
update instead of removing it immediately after an index is ready
for searches.

Closes ES-10697

idegtiarenko committed:
replace tuples with named parameters in AggregateMapper (#121542)

tlrx committed:
Use InMemoryNoOpCommitDirectory for archives indices only (#121210)

Since #118606 searchable snapshots shards are not expected to 
write files on disk, with the exception of archives indices mounted 
as searchable snapshots which require to rewrite the segment infos 
file in a newer version.

Ideally we should be able to remove the usage of the 
InMemoryNoOpCommitDirectory for non-archives searchable 
snapshots indices and only rely on SearchableSnapshotDirectory 
that throws on write operations. Similarly, starting 9.0 searchable 
snapshots shards do not write files on disk and therefore should 
be able to use a Directory implementation that forbids writes. 
Searchable snapshots shards for indices created before 9.0 
require a mutable directory for peer-recoveries.

In this change, we only allow writes for archives indices and 
searchable snapshots created before 9.0.

Relates ES-10438

gmarouli committed:
`DeprecationInfoAction` refactoring (#121181)

This refactoring was motivated by the following issues with the current
state of the code:

- The `TransformDeprecationChecker` is listed as plugin checker, but later we remove is from the `plugin_settings` and add it to the `cluster_settings`. This made me consider that the checker might be dealing with transform deprecation warnings but if they are listed under the `cliuster_settings`, it fits better to be part of `ClusterDeprecationChecker`.
- The `DeprecationInfo` is a data class, but it has a method `from` which constructs an `DeprecationInfo.Response` instance. However, this is not a simple factory class but it actually runs all the checks and it also tries to assert that it is not executed on a transport thread. Considering this, I thought it might fit better to the `TransportDeprecationInfoAction`, this way all the logic is in one place and all the checkers are wired and used in the same class.
- Constructing the node settings deprecation issues requires to merge the deprecation warnings of the individual nodes. We considered bringing together the execution of the remote request and the construction of the response in a new class called `NodeDeprecationChecker` that resembles the patterns of the other Checker classes.
- Reinstated the `PLUGIN_CHECKERS` even if we have only one check, so other developers can easier add their plugin checks.
- Finally, we noticed that the way we synthesise the remote requests is difficult to read and maintain because each call is nested under the previous one. We propose in this PR a different pattern that uses the `RefCountingListener` to combine the different remote calls and store their results in a container class named `PrecomputedData`
- **Bonus**: Removed the `LegacyIndexTemplateDeprecationChecker.java` which was not used.

elasticsearchmachine committed:
Mute org.elasticsearch.upgrades.UpgradeClusterClientYamlTestSuiteIT test {p0=mixed_cluster/110_enrich/Enrich stats query smoke test for mixed cluster} #121642

pxsalehi committed:
Fix AllocationFailuresResetOnShutdownIT#testResetAllocationFailuresOnNodeShutdown (#121378)

The ITs assumes that `ensureRed` means we've exhausted all the attempts for allocation, which doesn't hold always. 

Closes https://github.com/elastic/elasticsearch/issues/121129

cbuescher committed:
Add migration test for .async-search system index (#121517)

This change adds migration testing for the ".async-search" system index
to the full cluster upgrade tests that perform updates from versions N-2
to N via N-1.
The test creates a system index by using async_search on a cluster with
version N-2, then calls the "_migrate" API in version N-1 and finally
checks that on the upgraded cluster in N we are still able to retrieve
async search results from previous versions and can still write to the
system index. This is necessary to ensure we don't end up with a
write-only async search system index when migrating to version 9.

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/voyageai/VoyageAIEmbeddingsRequestEntity.java:
```diff
@@ -0,0 +1,81 @@
+        builder.field(INPUT_FIELD, input);
+        builder.field(MODEL_FIELD, model);
+
+        if (taskSettings.getInputType() != null) {
+            builder.field(INPUT_TYPE_FIELD, convertToString(taskSettings.getInputType()));
+        }
```
```suggestion
        var inputType = convertToString(taskSettings.getInputType());
        if (inputType != null) {
            builder.field(INPUT_TYPE_FIELD, inputType);
        }
```

Because `convertToString()` may return `null`

luigidellaquila committed:
Unmute docs.testFilterToday JDBC test (#121509)

Fixes: https://github.com/elastic/elasticsearch/issues/121474

Fixed by https://github.com/elastic/elasticsearch/pull/121504

ldematte committed:
[Entitlements] Fix Entitlement initialization to work across multiple versions (#121192)

slobodanadamovic committed:
Disable queryable built-in roles feature for core and datastream YAML tests (#121541)

This PR disables the "queryable built-in roles" feature for the `CoreWithSecurityClientYamlTestSuiteIT` and `DataStreamsClientYamlTestSuiteIT` YAML test suites. 
The feature was enabled by default in the #120323 PR, which asynchronously creates the `.security` index after cluster formation and indexes all built-in roles. The asynchronous creation of the `.security` index introduces non-deterministic behavior in our YAML tests. 
Since these test suites are not intended to verify the queryable built-in roles functionality, having the feature enabled introduced flakiness and unnecessary complexity to handle `.security` in existing tests. These tests would have to exclude the `.security` index in some way (by adjusting permissions or API calls), and in the end cleanup (delete) the `.security` index. This simply adds overhead without much gain. The feature is already test covered by `XPackRestIT` and other integration/REST tests, disabling it here would not compromise test coverage. Instead, it ensures these suites remain deterministic and focused on the behaviors they were designed to verify.

Resolves https://github.com/elastic/elasticsearch/issues/121536 
Resolves https://github.com/elastic/elasticsearch/issues/121513 
Resolves https://github.com/elastic/elasticsearch/issues/121484 
Resolves https://github.com/elastic/elasticsearch/issues/121478 
Resolves https://github.com/elastic/elasticsearch/issues/121290 
Resolves https://github.com/elastic/elasticsearch/issues/121246 
Resolves https://github.com/elastic/elasticsearch/issues/121242 
Resolves https://github.com/elastic/elasticsearch/issues/121238 
Resolves https://github.com/elastic/elasticsearch/issues/121186 
Resolves https://github.com/elastic/elasticsearch/issues/121131 
Resolves https://github.com/elastic/elasticsearch/issues/121130 
Resolves https://github.com/elastic/elasticsearch/issues/121128 
Resolves https://github.com/elastic/elasticsearch/issues/121014 
Resolves https://github.com/elastic/elasticsearch/issues/120965 
Resolves https://github.com/elastic/elasticsearch/issues/120920 
Resolves https://github.com/elastic/elasticsearch/issues/120890

rjernst committed:
Mute org.elasticsearch.search.CrossClusterSearchUnavailableClusterIT.testSearchSkipUnavailable

relates https://github.com/elastic/elasticsearch/issues/121497

ivancea committed:
ESQL: Fix AggregatorFunctionSupplierImplementer warnings import (#121654)

Extracted from https://github.com/elastic/elasticsearch/pull/116170/files#diff-d8dcb5e856b601ac28b3dc0055e09d5c815d2577705eade446bd55b2fca42ac7R145

Those warnings are not being used in main yet, so there's no error, and that code wasn't being executed. But it better be fixed in main already

jonathan-buttner committed:
[ML] Remove tasktype any from supportedStreamingTasks (#121460)

* Refactoring supported streaming functionality

* Moving always retrying handler to the tests

* Fixing comment

timgrein committed:
[Inference API] Fix node local rate limit calculator tests for non-snapshot builds (#121527)

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/voyageai/embeddings/VoyageAIEmbeddingType.java:
```diff
@@ -0,0 +1,121 @@
+        return elementType;
+    }
+
+    /**
+     * Returns an embedding type that is known based on the transport version provided. If the embedding type enum was not yet
+     * introduced it will be defaulted INT8.
+     *
+     * @param embeddingType the value to translate if necessary
+     * @param version the version that dictates the translation
+     * @return the embedding type that is known to the version passed in
+     */
+    public static VoyageAIEmbeddingType translateToVersion(VoyageAIEmbeddingType embeddingType, TransportVersion version) {
+        if (version.before(TransportVersions.V_8_14_0) && embeddingType == BYTE) {
+            return INT8;
+        }
+
+        return embeddingType;
+    }
```
```suggestion
```

Voyage AI won't have these backwards compatibility issues

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.action.CrossClusterAsyncQueryStopIT testStopQueryLocal #121672

masseyke committed:
Fail the reindex data stream task if any document fails to reindex (#121591)

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/voyageai/embeddings/VoyageAIEmbeddingsTaskSettings.java:
```diff
@@ -0,0 +1,249 @@
+        return requestTaskSettings.getTruncation() == null ? originalSettings.truncation : requestTaskSettings.getTruncation();
+    }
+
+    private static Integer getValidOutputDimension(
```
The difference between ServiceSettings and TaskSettings probably isn't very clean but the idea is that ServiceSettings cannot be changed and TaskSettings can change and can be overridden in the `POST _inference` API call. The reason ServiceSettings cannot change is to ensure compatibility. For example. if you change the model used to create a text embedding or change the `embedding_type` the resulting embeddings won't be comparable with embeddings created with another model or embedding type. 

The number of dimensions should be fixed and defined in ServiceSettings. When Elasticsearch creates a HNSW graph to store the dense vectors all the vectors must have the same number of dimensions, this is the `dims` parameter in the [Dense Vector field mappings](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html#dense-vector-params). Changing the number of dimensions breaks indexing new documents which is why we put the dimensions option in the immutable ServiceSettings.

fzowl replied:
@davidkyle Thank you very much or the explanation, it helped a lot!

davidkyle reviewed (commented):
This looks great. The PR is missing tests but once those are added please create the proper PR in the Elasticsearch repo

cbuescher committed:
Fix rare failures in YAML xContent roundtrip tests (#121515)

Under very unfortunate conditions tests that check xContent objects
roundtrip parsing  (like i.e. SearchHitsTests#testFromXContent)
can fail when we happen to randomly pick YAML xContent type and create
random (realistic)Unicode character sequences that may contain the
character U+0085 (133) from the Latin1 code page. That specific character
doesn't get parsed back to its original form for YAML xContent, which can 
lead to rare but hard to diagnose test failures.

This change adds logic to AbstractXContentTestCase#test() which lies at
the core of most of our  xContent roundtrip tests that disallows test
instances containing that particular character  when using YAML xContent
type.

Closes #97716

fang-xing-esql committed:
[ES|QL] Change function_named_parameters in Kibana doc to expected format (#121585)

* change function_named_parameters in kibana doc to expected format

rjernst committed:
re-mute HeapAttackIT.testLookupExplosionBigStringManyMatches

relates https://github.com/elastic/elasticsearch/issues/121465

dnhatn committed:
Clean up exchanges in EsqlNodeFailureIT (#121633)

If the query hits the failing index first, we will cancel the request,
preventing exchange-sink requests and data-node requests from reaching
another data node. As a result, exchange sinks could stay for 30
seconds.

fzowl committed:
Adding initial tests
Moving dimensions to ServiceSettings
